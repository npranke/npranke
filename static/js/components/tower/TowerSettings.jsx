import PropTypes from 'prop-types'
import React from 'react'

import useTimeRunner from '@components/hooks/TimeRunner'

function TowerSettings(props) {
    const { centiseconds, seconds, minutes } = useTimeRunner(
        props.isTimeRunning,
        props.shouldResetTime,
    )

    const twoDisks = React.useRef(null)
    const threeDisks = React.useRef(null)
    const fourDisks = React.useRef(null)
    const fiveDisks = React.useRef(null)

    function disksInputChangeHandler(event) {
        props.updateDisks(event.target.value)
    }

    function disksLabelKeyUpHandler(event) {
        const { value } = event.currentTarget.dataset

        if (event.key === 'Enter' || event.key === ' ') {
            props.updateDisks(value)
        } else if (event.key === 'ArrowRight') {
            if (value === '2') {
                threeDisks.current.focus()
            } else if (value === '3') {
                fourDisks.current.focus()
            } else if (value === '4') {
                fiveDisks.current.focus()
            } else if (value === '5') {
                twoDisks.current.focus()
            }
        } else if (event.key === 'ArrowLeft') {
            if (value === '2') {
                fiveDisks.current.focus()
            } else if (value === '3') {
                twoDisks.current.focus()
            } else if (value === '4') {
                threeDisks.current.focus()
            } else if (value === '5') {
                fourDisks.current.focus()
            }
        }
    }

    return (
        <div className="settings">
            <div className="settings-inner">
                <div className="left-text">
                    <span className="disks-text">
                        <span className="disks-portion" id="disks-label">
                            disks:&ensp;
                        </span>
                        <span
                            className="disks-portion"
                            id="disks-radiogroup"
                            role="radiogroup"
                            aria-labelledby="disks-label"
                        >
                            <input
                                name="disks"
                                type="radio"
                                id="two-disks"
                                value="2"
                                onChange={ disksInputChangeHandler }
                                checked={ props.disks === 2 }
                                aria-checked={ props.disks === 2 }
                            />
                            <label
                                className="disks-portion number"
                                id="two-disks-label"
                                htmlFor="two-disks"
                                data-value="2"
                                ref={ twoDisks }
                                onKeyUp={ disksLabelKeyUpHandler }
                                tabIndex={ props.disks === 2 ? '0' : '-1' }
                            >
                                2
                            </label>
                            <input
                                name="disks"
                                type="radio"
                                id="three-disks"
                                value="3"
                                onChange={ disksInputChangeHandler }
                                checked={ props.disks === 3 }
                                aria-checked={ props.disks === 3 }
                            />
                            <label
                                className="disks-portion number"
                                id="three-disks-label"
                                htmlFor="three-disks"
                                data-value="3"
                                ref={ threeDisks }
                                onKeyUp={ disksLabelKeyUpHandler }
                                tabIndex={ props.disks === 3 ? '0' : '-1' }
                            >
                                3
                            </label>
                            <input
                                name="disks"
                                type="radio"
                                id="four-disks"
                                value="4"
                                onChange={ disksInputChangeHandler }
                                checked={ props.disks === 4 }
                                aria-checked={ props.disks === 4 }
                            />
                            <label
                                className="disks-portion number"
                                id="four-disks-label"
                                htmlFor="four-disks"
                                data-value="4"
                                ref={ fourDisks }
                                onKeyUp={ disksLabelKeyUpHandler }
                                tabIndex={ props.disks === 4 ? '0' : '-1' }
                            >
                                4
                            </label>
                            <input
                                name="disks"
                                type="radio"
                                id="five-disks"
                                value="5"
                                onChange={ disksInputChangeHandler }
                                checked={ props.disks === 5 }
                                aria-checked={ props.disks === 5 }
                            />
                            <label
                                className="disks-portion number"
                                id="five-disks-label"
                                htmlFor="five-disks"
                                data-value="5"
                                ref={ fiveDisks }
                                onKeyUp={ disksLabelKeyUpHandler }
                                tabIndex={ props.disks === 5 ? '0' : '-1' }
                            >
                                5
                            </label>
                        </span>
                    </span>
                </div>
                <div className="right-text">
                    <span className="moves-text">
                        <span className="moves-portion">
                            moves:&ensp;
                        </span>
                        <span className="moves-portion number">
                            {
                                props.moves < 10
                                    ? `0${props.moves}`
                                    : `${props.moves}`
                            }
                        </span>
                    </span>
                    <span className="time-text">
                        <span className="time-portion">
                            time:&emsp;
                        </span>
                        <span className="time-portion number">
                            { minutes }
                        </span>
                        <span className="time-portion">
                            &ensp;:&ensp;
                        </span>
                        <span className="time-portion number">
                            { seconds }
                        </span>
                        <span className="time-portion">
                            &ensp;:&ensp;
                        </span>
                        <span className="time-portion number">
                            { centiseconds }
                        </span>
                    </span>
                </div>
            </div>
        </div>
    )
}

TowerSettings.defaultProps = {
    disks: 3,
    isTimeRunning: false,
    shouldResetTime: false,
    moves: 0,
}

TowerSettings.propTypes = {
    disks: PropTypes.number,
    isTimeRunning: PropTypes.bool,
    shouldResetTime: PropTypes.bool,
    moves: PropTypes.number,
    updateDisks: PropTypes.func.isRequired,
}

export default TowerSettings
