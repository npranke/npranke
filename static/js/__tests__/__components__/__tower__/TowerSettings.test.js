import Adapter from 'enzyme-adapter-react-16'
import Enzyme, { mount, shallow } from 'enzyme'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import TowerSettings from '@components/tower/TowerSettings'

Enzyme.configure({ adapter: new Adapter() })

describe('TowerSettings', () => {
    let mockUpdateDisks

    beforeAll(() => {
        mockUpdateDisks = jest.fn()
    })

    afterEach(() => {
        mockUpdateDisks.mockReset()
    })

    test('has disks-portion sections', () => {
        const towerSettings = shallow(
            <TowerSettings updateDisks={ jest.fn() } />,
        )

        expect(
            towerSettings.find('.disks-portion'),
        ).toHaveLength(6)
    })

    test('has disks-portion number sections', () => {
        const towerSettings = shallow(
            <TowerSettings updateDisks={ jest.fn() } />,
        )

        expect(
            towerSettings.find('.disks-portion.number'),
        ).toHaveLength(4)
    })

    test('shows moves number when less than 10', () => {
        const towerSettings = shallow(
            <TowerSettings moves={ 1 } updateDisks={ jest.fn() } />,
        )

        expect(
            towerSettings.find('.moves-portion.number').text(),
        ).toEqual('01')
    })

    test('shows moves number when 10', () => {
        const towerSettings = shallow(
            <TowerSettings moves={ 10 } updateDisks={ jest.fn() } />,
        )

        expect(
            towerSettings.find('.moves-portion.number').text(),
        ).toEqual('10')
    })

    test('shows moves number when more than 10', () => {
        const towerSettings = shallow(
            <TowerSettings moves={ 15 } updateDisks={ jest.fn() } />,
        )

        expect(
            towerSettings.find('.moves-portion.number').text(),
        ).toEqual('15')
    })

    test('has time-portion sections', () => {
        const towerSettings = shallow(
            <TowerSettings updateDisks={ jest.fn() } />,
        )

        expect(
            towerSettings.find('.time-portion'),
        ).toHaveLength(6)
    })

    test('has time-portion number sections', () => {
        const towerSettings = shallow(
            <TowerSettings updateDisks={ jest.fn() } />,
        )

        expect(
            towerSettings.find('.time-portion.number'),
        ).toHaveLength(3)
    })

    describe('disks selection radiogroup', () => {
        test('has one radiogroup', () => {
            const towerSettings = shallow(
                <TowerSettings updateDisks={ jest.fn() } />,
            )

            expect(
                towerSettings.find('#disks-radiogroup').props().role,
            ).toEqual('radiogroup')
        })

        test('has four radio inputs', () => {
            const towerSettings = shallow(
                <TowerSettings updateDisks={ jest.fn() } />,
            )

            expect(
                towerSettings.find('#disks-radiogroup input'),
            ).toHaveLength(4)

            towerSettings.find('#disks-radiogroup input').forEach(
                (input) => {
                    expect(input.props().type).toEqual('radio')
                },
            )
        })

        test('has four input labels', () => {
            const towerSettings = shallow(
                <TowerSettings updateDisks={ jest.fn() } />,
            )

            expect(
                towerSettings.find('#disks-radiogroup label'),
            ).toHaveLength(4)

            towerSettings.find('#disks-radiogroup label').forEach(
                (label) => {
                    expect(label.props().htmlFor).toBeTruthy()
                },
            )
        })

        test('disksInputChangeHandler() calls updateDisks', async () => {
            const user = userEvent.setup()

            const { rerender } = render(
                <TowerSettings
                    disks={ 3 }
                    updateDisks={ mockUpdateDisks }
                />,
            )
            await user.click(screen.getByRole('radio', { name: '2' }))

            expect(mockUpdateDisks).toHaveBeenCalledTimes(1)
            expect(mockUpdateDisks).toHaveBeenLastCalledWith('2')

            rerender(
                <TowerSettings disks={ 2 } updateDisks={ mockUpdateDisks } />,
            )
            await user.click(screen.getByRole('radio', { name: '3' }))

            expect(mockUpdateDisks).toHaveBeenCalledTimes(2)
            expect(mockUpdateDisks).toHaveBeenLastCalledWith('3')

            rerender(
                <TowerSettings disks={ 3 } updateDisks={ mockUpdateDisks } />,
            )
            await user.click(screen.getByRole('radio', { name: '4' }))

            expect(mockUpdateDisks).toHaveBeenCalledTimes(3)
            expect(mockUpdateDisks).toHaveBeenLastCalledWith('4')

            rerender(
                <TowerSettings disks={ 4 } updateDisks={ mockUpdateDisks } />,
            )
            await user.click(screen.getByRole('radio', { name: '5' }))

            expect(mockUpdateDisks).toHaveBeenCalledTimes(4)
            expect(mockUpdateDisks).toHaveBeenLastCalledWith('5')
        })

        describe('disksLabelKeyUpHandler()', () => {
            beforeEach(() => {
                document.body.innerHTML = '<div id="fastener"></div>'
            })

            test('calls updateDisks prop on enter', () => {
                const towerSettings = mount(
                    <TowerSettings updateDisks={ jest.fn() } />,
                )

                towerSettings.find('#five-disks-label').simulate(
                    'keyup',
                    { key: 'Enter' },
                )

                expect(
                    towerSettings.props().updateDisks,
                ).toHaveBeenCalledWith('5')
            })

            test('calls updateDisks prop on spacebar', () => {
                const towerSettings = mount(
                    <TowerSettings updateDisks={ jest.fn() } />,
                )

                towerSettings.find('#three-disks-label').simulate(
                    'keyup',
                    { key: ' ' },
                )

                expect(
                    towerSettings.props().updateDisks,
                ).toHaveBeenCalledWith('3')
            })

            test('updates active element for arrow right on two', () => {
                const towerSettings = mount(
                    <TowerSettings updateDisks={ jest.fn() } />,
                    { attachTo: fastener },
                )

                towerSettings.find('#two-disks-label').simulate(
                    'keyup',
                    { key: 'ArrowRight' },
                )

                expect(
                    document.activeElement.id,
                ).toEqual('three-disks-label')
            })

            test('updates active element for arrow right on three', () => {
                const towerSettings = mount(
                    <TowerSettings updateDisks={ jest.fn() } />,
                    { attachTo: fastener },
                )

                towerSettings.find('#three-disks-label').simulate(
                    'keyup',
                    { key: 'ArrowRight' },
                )

                expect(
                    document.activeElement.id,
                ).toEqual('four-disks-label')
            })

            test('updates active element for arrow right on four', () => {
                const towerSettings = mount(
                    <TowerSettings updateDisks={ jest.fn() } />,
                    { attachTo: fastener },
                )

                towerSettings.find('#four-disks-label').simulate(
                    'keyup',
                    { key: 'ArrowRight' },
                )

                expect(
                    document.activeElement.id,
                ).toEqual('five-disks-label')
            })

            test('updates active element for arrow right on five', () => {
                const towerSettings = mount(
                    <TowerSettings updateDisks={ jest.fn() } />,
                    { attachTo: fastener },
                )

                towerSettings.find('#five-disks-label').simulate(
                    'keyup',
                    { key: 'ArrowRight' },
                )

                expect(
                    document.activeElement.id,
                ).toEqual('two-disks-label')
            })

            test('updates active element for arrow left on two', () => {
                const towerSettings = mount(
                    <TowerSettings updateDisks={ jest.fn() } />,
                    { attachTo: fastener },
                )

                towerSettings.find('#two-disks-label').simulate(
                    'keyup',
                    { key: 'ArrowLeft' },
                )

                expect(
                    document.activeElement.id,
                ).toEqual('five-disks-label')
            })

            test('updates active element for arrow left on three', () => {
                const towerSettings = mount(
                    <TowerSettings updateDisks={ jest.fn() } />,
                    { attachTo: fastener },
                )

                towerSettings.find('#three-disks-label').simulate(
                    'keyup',
                    { key: 'ArrowLeft' },
                )

                expect(
                    document.activeElement.id,
                ).toEqual('two-disks-label')
            })

            test('updates active element for arrow left on four', () => {
                const towerSettings = mount(
                    <TowerSettings updateDisks={ jest.fn() } />,
                    { attachTo: fastener },
                )

                towerSettings.find('#four-disks-label').simulate(
                    'keyup',
                    { key: 'ArrowLeft' },
                )

                expect(
                    document.activeElement.id,
                ).toEqual('three-disks-label')
            })

            test('updates active element for arrow left on five', () => {
                const towerSettings = mount(
                    <TowerSettings updateDisks={ jest.fn() } />,
                    { attachTo: fastener },
                )

                towerSettings.find('#five-disks-label').simulate(
                    'keyup',
                    { key: 'ArrowLeft' },
                )

                expect(
                    document.activeElement.id,
                ).toEqual('four-disks-label')
            })
        })

        describe('input checked', () => {
            test('selected disks input has checked true', () => {
                const towerSettings = shallow(
                    <TowerSettings disks={ 2 } updateDisks={ jest.fn() } />,
                )

                expect(
                    towerSettings.find('#two-disks').props().checked,
                ).toEqual(true)
            })

            test('unselected disks inputs have checked false', () => {
                const towerSettings = shallow(
                    <TowerSettings disks={ 2 } updateDisks={ jest.fn() } />,
                )

                expect(
                    towerSettings.find('#three-disks').props().checked,
                ).toEqual(false)
                expect(
                    towerSettings.find('#four-disks').props().checked,
                ).toEqual(false)
                expect(
                    towerSettings.find('#five-disks').props().checked,
                ).toEqual(false)
            })
        })

        describe('label tabindex', () => {
            test('selected disks label has tabindex 0', () => {
                const towerSettings = shallow(
                    <TowerSettings disks={ 3 } updateDisks={ jest.fn() } />,
                )

                expect(
                    towerSettings.find('#three-disks-label').props().tabIndex,
                ).toEqual('0')
            })

            test('unselected disks labels have tabindex -1', () => {
                const towerSettings = shallow(
                    <TowerSettings disks={ 3 } updateDisks={ jest.fn() } />,
                )

                expect(
                    towerSettings.find('#two-disks-label').props().tabIndex,
                ).toEqual('-1')
                expect(
                    towerSettings.find('#four-disks-label').props().tabIndex,
                ).toEqual('-1')
                expect(
                    towerSettings.find('#five-disks-label').props().tabIndex,
                ).toEqual('-1')
            })
        })
    })
})

describe('TowerSettings snapshot', () => {
    test('matches snapshot', () => {
        const { asFragment } = render(
            <TowerSettings updateDisks={ jest.fn() } />,
        )

        expect(asFragment()).toMatchSnapshot()
    })
})
