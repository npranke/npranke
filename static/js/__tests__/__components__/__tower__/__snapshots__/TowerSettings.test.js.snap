// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TowerSettings snapshot matches snapshot 1`] = `
<div
  className="settings"
>
  <div
    className="settings-inner"
  >
    <div
      className="left-text"
    >
      <span
        className="discs-text"
      >
        <span
          className="discs-portion"
          id="discs-label"
        >
          discs: 
        </span>
        <span
          aria-labelledby="discs-label"
          className="discs-portion"
          id="discs-radiogroup"
          role="radiogroup"
        >
          <input
            aria-checked={false}
            checked={false}
            id="two-discs"
            name="discs"
            onChange={[Function]}
            type="radio"
            value="2"
          />
          <label
            className="discs-portion number"
            data-value="2"
            htmlFor="two-discs"
            id="two-discs-label"
            onKeyUp={[Function]}
            tabIndex="-1"
          >
            2
          </label>
          <input
            aria-checked={true}
            checked={true}
            id="three-discs"
            name="discs"
            onChange={[Function]}
            type="radio"
            value="3"
          />
          <label
            className="discs-portion number"
            data-value="3"
            htmlFor="three-discs"
            id="three-discs-label"
            onKeyUp={[Function]}
            tabIndex="0"
          >
            3
          </label>
          <input
            aria-checked={false}
            checked={false}
            id="four-discs"
            name="discs"
            onChange={[Function]}
            type="radio"
            value="4"
          />
          <label
            className="discs-portion number"
            data-value="4"
            htmlFor="four-discs"
            id="four-discs-label"
            onKeyUp={[Function]}
            tabIndex="-1"
          >
            4
          </label>
          <input
            aria-checked={false}
            checked={false}
            id="five-discs"
            name="discs"
            onChange={[Function]}
            type="radio"
            value="5"
          />
          <label
            className="discs-portion number"
            data-value="5"
            htmlFor="five-discs"
            id="five-discs-label"
            onKeyUp={[Function]}
            tabIndex="-1"
          >
            5
          </label>
        </span>
      </span>
    </div>
    <div
      className="right-text"
    >
      <span
        className="moves-text"
      >
        <span
          className="moves-portion"
        >
          moves: 
        </span>
        <span
          className="moves-portion number"
        >
          00
        </span>
      </span>
      <span
        className="time-text"
      >
        <span
          className="time-portion"
        >
          time: 
        </span>
        <span
          className="time-portion number"
        >
          00
        </span>
        <span
          className="time-portion"
        >
           : 
        </span>
        <span
          className="time-portion number"
        >
          00
        </span>
        <span
          className="time-portion"
        >
           : 
        </span>
        <span
          className="time-portion number"
        >
          00
        </span>
      </span>
    </div>
  </div>
</div>
`;
